PWGJE/Tasks/bjetTreeCreator.cxx

```diff
@@ -0,0 +1,634 @@
+    fastjet::PseudoJet parentSubJet1;
+    fastjet::PseudoJet parentSubJet2;
+    bool softDropped = false;
+    double nSD = 0.0;
+    double Zg = -1.0;
+    double Rg = -1.0;
+    while (daughterSubJet.has_parents(parentSubJet1, parentSubJet2)) {
+      if (parentSubJet1.perp() < parentSubJet2.perp()) {
+        std::swap(parentSubJet1, parentSubJet2);
+      }
+
+      // delta_R_vec.push_back(delta_R);
+      // xkt_vec.push_back(xkt);
+      // z_vec.push_back(z);
+
+      auto z = parentSubJet2.perp() / (parentSubJet1.perp() + parentSubJet2.perp());
+      auto theta = parentSubJet1.delta_R(parentSubJet2); // TODO add the lund plane
+      // double xkt = parentSubJet2.perp() * sin(theta);  // TODO add the lund plane
+      if (z >= zCut * TMath::Power(theta / (jet.r() / 100.f), beta)) {
+        if (!softDropped) {
+          Zg = z;
+          Rg = theta;
+          softDropped = true;
+        }
+        nSD++;
+      }
+      daughterSubJet = parentSubJet1;
+    }
+    return std::make_tuple(nSD, Zg, Rg);
+  }
+
+  template <typename AnyJet, typename AllMCParticles>
+  short getJetFlavor(AnyJet const& jet, AllMCParticles const& mcparticles)
+  {
+    const int arraySize = 99;
+
+    std::array<int, arraySize> countpartcode;
+    int count = 0;
+
+    for (auto& mcpart : mcparticles) {
+      int pdgcode = mcpart.pdgCode();
+      if (TMath::Abs(pdgcode) == 21 || (TMath::Abs(pdgcode) >= 1 && TMath::Abs(pdgcode) <= 5)) {
+        double dR = jetutilities::deltaR(jet, mcpart);
+
+        if (dR < jet.r() / 100.f) {
+          if (TMath::Abs(pdgcode) == 5) {
+            return 2; // Beauty jet
+          } else {
+            if (count > arraySize - 1)
+              return 0;
+            countpartcode[count] = pdgcode;
+            count++;
+          }
+        }
+      }
+    }
+
+    for (int ij = 0; ij < count; ij++) {
+      if (TMath::Abs(countpartcode[ij]) == 4)
+        return 1; // Charm jet
+    }
+
+    return 0; // light flavor jet
+  }
+
+  // NOTE: The prongs of the SVs are double counted in different SVs, see if this has any effect !!!
+  template <typename AnalysisJet, typename SecondaryVertices>
+  void processJetSVInfo(AnalysisJet const& myJet, SecondaryVertices const& allSVs, std::vector<int>& svIndices, int jetFlavor = 0, double eventweight = 1.0)
+  {
+    for (const auto& candSV : myJet.template secondaryVertices_as<SecondaryVertices>()) {
+
+      if (candSV.pt() < svPtMin) {
+        continue;
+      }
+
+      double deltaRJetSV = jetutilities::deltaR(myJet, candSV);
+      double massSV = candSV.m();
+      double energySV = candSV.e();
+
+      if (produceTree) {
+        bjetSVParamsTable(bjetParamsTable.lastIndex() + 1, candSV.pt(), deltaRJetSV, massSV, energySV / myJet.energy(), candSV.impactParameterXY(), candSV.cpa(), candSV.chi2PCA(), candSV.decayLengthXY(), candSV.errorDecayLengthXY(), candSV.decayLength(), candSV.errorDecayLength());
+      }
+      svIndices.push_back(bjetSVParamsTable.lastIndex());
```

Comment:
what is this cheking? the number of SV per jet? Cant you just do jet.secondaryVertices.size()?

Reply:
No. This is not just for the secondary vertex numbers. This is an index table will be used to connect the jet table to the SV table.